### ISSUE TITLE
DataLoader batching doesn't work properly with lists

### ISSUE THREAD

**Issue by arlampin** [graphql-java/graphql-java#831]
Hello,

There seems to be still some batching issues with DataLoader and DataLoaderDispatcherInstrumentation, even after #764 was merged.

I created a simple example in https://gist.github.com/arlampin/183d40ab7e8eb2507917fe8c46730189 that demonstrates the issue.

In the example, there is a simple tree of 7 items, like so:
```
      1
    /   \
   2     3
  / \   / \
 4   5 6   7
```

Using following query `query Q { root { id childNodes { id  childNodes {  id  childNodes { id }}}}}` (where 'root' is node 1) results in output:
```
BatchLoader called for [1] -> got [[2, 3]]
BatchLoader called for [2, 3] -> got [[4, 5], [6, 7]]
BatchLoader called for [4, 5] -> got [[], []]
BatchLoader called for [6, 7] -> got [[], []]
```
If I've understood correctly the idea behind DataLoader, there should only be three calls, one for each level in tree, and calls to [4, 5] and [6, 7] should have been merged into a single call.

This example doesn't actually do any true async execution, but replacing `CompletableFuture.completedFuture` with an async call doesn't affect the results.

**Comment by bbakerman** [comment]
I have looked further into this and we have structural problems in the ExecutionStrategy that means that we dont descend in perfect sets of layers

So given your tree

```

      1
    /   \
   2    3
  / \   / \
4  5 6  7
```

The code does this per field

```
        CompletableFuture<ExecutionResult> result = fetchField(executionContext, parameters)
                .thenCompose((fetchedValue) ->
                        completeField(executionContext, parameters, fetchedValue));
```

That is fetch the field (which can involve the dataloader) and complete the field which can cause the engine to descend and restart itself recursively.

There for the order can be as you observed.  As it gets to node `2` it descends its list because it calls `dispatch` for that inner recursive call.

I think we would need to rewrite the ExecutionStrategy so that fetched all fields first in a level and then called resolve.  But even then it would need to know about `lists` and `object types` since they cause the engine to descend for those fields.

Hmmmm

**Comment by samogot** [comment]
Even if you rewrite ExecutionStrategy such way, this will work only for first layer. After recursion call you won't be able to know if you are in recursive call and don't need to dispatch, or it is last recursion branch of a layer an you need run dispatch now
I had an idea to rewrite `DataLoaderDispatcherInstrumentation` to do something like reference counting of future recursive calls, and run dispatch only once in `end` `CompleteFields`. But I gave this idea up, because Instrumentation do not have all data that I need (`ExecutionContext`) and because this is very error-prone strategy and needs a lot of testing

**Comment by arlampin** [comment]
Are there any plans for improving DataLoader batching regarding this issue? We are currently using BatchedExecutionStrategy because it handles this case properly, but would like to switch to async DataLoaders.

I wonder if BatchedExecutionStrategy itself contains something that could be reused for this?

**Comment by vojtechhabarta** [comment]
I have the same problem.
Currently I am using `BatchedExecutionStrategy` and I wanted to switch to `AsyncExecutionStrategy` but it is much slower in my case and for larger data (where `BatchedExecutionStrategy` finishes its job in 30 seconds) `AsyncExecutionStrategy` fails after few minutes with `OutOfMemoryError`.

I described my use case here https://github.com/graphql-java/graphql-java/issues/760#issuecomment-372826395 since I overlooked this issue.

**Comment by bbakerman** [comment]
We have greatly improved the data loader performance by "tracking" the levels encountered and dispatching the DL when we know we have have "dispatched" all the outstanding field fetchers including ones for lists.

We have tests that have shown it reduces it greatly.

I am going to close this issue as the  9.0 release is imminent

**Comment by bbakerman** [comment]
Whoops - jumped the gun - this didnt make it into 9.0.  Sorry

**Comment by bbakerman** [comment]
See https://github.com/graphql-java/graphql-java/pull/990

**Comment by arlampin** [comment]
I take it this made it to 9.0 after all?

**Comment by andimarek** [comment]
@arlampin yes, we actually managed to ship it in 9.0.

**Comment by andimarek** [comment]
closing it again, because it is fixed in 9.0.
